### PR TITLE
Feature plugin manager

### PR DESCRIPTION
**Pull request by dgreene-r7**
This is the first of two pull requests for the `PluginManager`. The module is complete, however test coverage only focuses on instantiation and error conditions on startup. The next pull request will focus on testing persistent behavior such as handling events from sources and failures during operation.

### CODE REVIEW

File: lib/plugin-manager.js [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+'use strict';
+
+const EventEmitter = require('events').EventEmitter;
+const async = require('async');
+const StringTemplate = require('./string-template');
+const Metadata = require('./source/metadata');
+const S3 = require('./source/s3');
+
+// const Consul = require('./source/consul');
+// const File = require('./source/file');
```
Comment (fmitchell-r7) [inline lib/plugin-manager.js]: Probably worth a "TODO" comment that these are in progress. Or just take out the commented lines.

File: lib/plugin-manager.js [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        value = callback(prop, obj[prop]);
+      }
+      collected[prop] = value;
+    }
+  }
+  return collected;
+}
+
+class PluginManager extends EventEmitter {
+  constructor(storage) {
```
Comment (fmitchell-r7) [inline lib/plugin-manager.js]: Do we want a doc comment here to indicate what type `storage` is?

File: lib/plugin-manager.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    this._running = false;
+    this._ok = true;
+    this.updateDelay = DEFAULT_CALLBACK_DELAY;
+    this._boundInit = this._init.bind(this);
+  }
+
+  /**
+   * Initialize the PluginManager's index and metadata source
+   */
+  init() {
```
Comment (fmitchell-r7) [inline lib/plugin-manager.js]: Should we name this `initialize` so it's consistent with plugins?

File: lib/plugin-manager.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    Log.info('Initializing index and metadata');
+    this._running = true;
+    this._init();
+  }
+
+  /**
+   * Shutdown the plugin manager's sources and then the plugin manager
+   * @param {Boolean} del
```
Comment (fmitchell-r7) [inline lib/plugin-manager.js]: What does the `del` parameter do?

File: lib/plugin-manager.js [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+   * Create function that feeds async.parallel (or any async marshalling function)
+   * @param {Metadata|S3} source
+   * @returns {Function}
+   * @private
+   */
+  _createParallelTask(source) {
+    return (c) => {
+      const callBackErr = (err) => {
+        source.removeListener('update', callBackSuccess); // eslint-disable-line no-use-before-define
+        c(err);
```
Comment (fmitchell-r7) [inline lib/plugin-manager.js]: Can you pass `source.type` into the callback here e.g. `c(err, source.type)`? That might help avoid the "which plugin had an error" question that came up in the `_index` function.

File: lib/plugin-manager.js [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+      this.storage.register(instance);
+      this._registerSourceEvents(instance);
+      this.emit('source-registered', instance);
+
+      instance.initialize();
+      this.emit('source-initialized', instance);
+    });
+
+    this.index.once('update', () => {
+      // TODO: Implement timeout so that someone spamming updates to the GH repo won't cause a constant rebuild
```
Comment (fmitchell-r7) [inline lib/plugin-manager.js]: Do you still need this TODO? It looks like you've got a working timeout.

File: lib/plugin-manager.js [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        this.storage.update();
+      }, this.updateDelay);
+    });
+
+    source.on('no-update', () => {
+      Log.info(`${source.name} has no update to its underlying source data.`);
+    });
+
+    source.on('error', (err) => {
+      Log.info(`${source.name} encountered the following error: ${err}`);
```
Comment (fmitchell-r7) [inline lib/plugin-manager.js]: Do we want to log this as `error` instead of `info`?

File: lib/plugin-manager.js (hunk 4 above)
Comment (dgreene-r7) [inline lib/plugin-manager.js]: The `del` parameter clears the `Storage` layer's sources array. It was mainly a proposal until we figured out a) how the `PluginManager` shutting down affects its `Storage` object and b) how the `Storage` layer handles clearing its sources.

File: lib/plugin-manager.js (hunk 5 above)
Comment (dgreene-r7) [inline lib/plugin-manager.js]: I think adding `source.type` to the callback is a great idea. I'm going to do some investigation to see if it's actually possible to pass additional arguments into the `err` case of `async.parallel`.

File: lib/plugin-manager.js [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+  /**
+   * Maps instance metadata into the index document
+   * @param {Function} callback
+   * @private
+   */
+  _mapper(callback) {
+    let sources = [];
+
+    try {
+      sources = this.index.properties.sources.map((el) => {
```
Comment (jmanero-r7) [inline lib/plugin-manager.js]: This isn't going to line up with the behaviors at https://github.com/rapid7/propsd/blob/master/lib/storage.js#L70 and https://github.com/rapid7/propsd/blob/master/lib/source/s3.js#L18 (which isn't parsing the document defined in https://github.com/rapid7/propsd/blob/master/docs/schemas.md#s3-stored-json-configuration-files correctly). `properties` and `sources` should be top-level keys in the document.

Comment (dgreene-r7) [inline lib/plugin-manager.js]: Can you comment on where the test cases are misconfigured? I copied the test data (in `test/data/s3/`) from the original test data you generated and all tests written against it are passing.

Comment (dgreene-r7) [inline lib/plugin-manager.js]: I should clarify that the index source isn't being registered with the storage layer by design. It's a special case with a different syntax from the rest of the S3 source documents.

**Comment by fmitchell-r7** [comment]
:+1: Looks good to me.

**Pull request merged**